### PR TITLE
mparser: add equality operators to nodes

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

**Comment by xclaesse** [comment]
This is extracted from https://github.com/mesonbuild/meson/pull/10151 as it's not directly related to cargo. I added an extra commit on top of @dcbaker's work to hopefully get CI to pass.

**Comment by xclaesse** [comment]
@dcbaker @eli-schwartz any objection?

File: mesonbuild/mparser.py [hunk 1]
```diff
@@ -240,24 +240,25 @@ def lex(self, filename: str) -> T.Generator[Token, None, None]:
...
-        if self.end_lineno is None:
-            self.end_lineno = self.lineno
-        if self.end_colno is None:
-            self.end_colno = self.colno
+    def __init__(self, lineno: int, colno: int, filename: str, end_lineno: T.Optional[int] = None, end_colno: T.Optional[int] = None) -> None:
+        self.lineno = lineno
+        self.colno = colno
+        self.filename = filename
+        self.end_lineno = end_lineno if end_lineno is not None else lineno
+        self.end_colno = end_colno if end_colno is not None else colno
```
Comment (dcbaker) [inline mesonbuild/mparser.py:256]: I think you can keep using the `__post_init__` here by doing something like:

```python
end_lineno: InitVar[T.Optional[int]] = field(default=None, hash=False)
end_colno: InitVar[T.Optional[int]] = field(default=None, hash=False)

def __post_init__(self, end_lineno: T.Optional[int], end_colno: T.Optional[int]):
    self.end_lineno = end_lineno if end_lineno is not None else self.lineno
    self.end_colno = end_colno if end_colno is not None else self.colno
```

But it's up to you which you prefer I don't have a strong opinion

**Review by dcbaker — APPROVED** [review]
I have a suggestion to leave the `__post_init__` in the `BaseNode`, which I kinda like because it's DRYer than the full `__init__` solution that you have, but I'm also not going to block this on that change. Either way this LGTM

File: mesonbuild/mparser.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/mparser.py:256]: mypy does not seems to like that solution. All other node classes have `__init__` too, so I don't think it's a big issue to keep it here.

Was that your only concern on this PR?

Comment (xclaesse) [reply mesonbuild/mparser.py:256]: oh, I see you approved the PR already, LGTM.

**Pull request merged**